### PR TITLE
Handle more places where package direcroy is converted to canonical file path

### PR DESCRIPTION
**Pull request by sheetalkamat**
1. Make sure package info is not queried on canonical file path
2. Verify packageDirectory before sending package info with same pacageDirectory from cache. This should ensure the packageJsonInfo should always have packageDirectory matching the case with requested path

Fixes: https://github.com/microsoft/TypeScript/pull/50557#issuecomment-1243452272

### CODE REVIEW

**Review by weswigham — APPROVED** [review]
This gets us consistent casing in the names we lookup, sure, so in that respect it's fine, but does it not also break using inconsistent name case on a case insensitive system (which otherwise should work)? Or are all of our comparison and caching sites already using case-appropriate comparisons? Our package caches in particular worry me - I don't think these caches use case-sensitive-appropriate key comparisons (but I could be remembering incorrectly), so I worry using non-path'd paths like this everywhere will result in duplicated cache entries on case-insensitve filesystems. Could be wrong though, might already be handled; but the eager path-ifying we were doing before makes me think it's not.

**Comment by sheetalkamat** [comment]
Cache key handles case sensitivity so that doesnt break this.

**Comment by sheetalkamat** [comment]
@typescript-bot cherry-pick this to release-4.8 and LKG

**Comment by sheetalkamat** [comment]
@typescript-bot cherry-pick this to release-4.8

File: src/compiler/moduleNameResolver.ts [hunk 1]
```diff
@@ -1818,7 +1822,7 @@ namespace ts {
      * A function for locating the package.json scope for a given path
      */
     /*@internal*/
-     export function getPackageScopeForPath(fileName: Path, state: ModuleResolutionState): PackageJsonInfo | undefined {
+     export function getPackageScopeForPath(fileName: string, state: ModuleResolutionState): PackageJsonInfo | undefined {
```
Comment (amcasey) [inline src/compiler/moduleNameResolver.ts:1825]: Personally, I would find "path" or "filePath" clearer.

File: src/compiler/moduleNameResolver.ts [hunk 2]
```diff
@@ -1845,7 +1849,9 @@ namespace ts {
             if (typeof existing !== "boolean") {
                 if (traceEnabled) trace(host, Diagnostics.File_0_exists_according_to_earlier_cached_lookups, packageJsonPath);
                 state.affectingLocations.push(packageJsonPath);
-                return existing;
+                return existing.packageDirectory === packageDirectory ?
```
Comment (amcasey) [inline src/compiler/moduleNameResolver.ts:1852]: This is a way to reuse the cache contents with a variant (e.g. casing, slashes) of the path?

File: src/compiler/moduleNameResolver.ts [hunk 3]
```diff
@@ -1808,6 +1808,10 @@ namespace ts {
     /*@internal*/
     export interface PackageJsonInfo {
         packageDirectory: string;
+        contents: PackageJsonInfoContents;
+    }
+    /*@internal*/
+    export interface PackageJsonInfoContents {
```
Comment (amcasey) [inline src/compiler/moduleNameResolver.ts:1814]: I guess we don't have so many package.json files that the extra allocations will be a problem?

**Review by amcasey — COMMENTED** [review]
Every time we touch the path code, I grow more convinced that it needs to be revamped.  We sure do create a lot of temporary strings walking up the directory spine...

**Pull request merged**